### DISCUSSION TITLE
How to use the DG kernel to construct the coupled equations?

### DISCUSSION THREAD

**Discussion by SomnusYu** [idaholab/moose#23079] [Q&A General] [answered]
Hi all,

I know the DG kernel can solve such formal equation conveniently:
![1672912578285](https://user-images.githubusercontent.com/118275988/210752396-e3b17bcd-4aea-4297-b995-f7966ccb27a5.png)
, where φ is the variable, _v_ is the velocity.

But how to construct the coupled equations? For example:
![1672912809819](https://user-images.githubusercontent.com/118275988/210753247-701d015f-3391-4978-8aec-36c50da505bb.png)
, where φ, θ are variables, _v_ is the velocity.

I don't know how to express this term (![1672912909530](https://user-images.githubusercontent.com/118275988/210753646-213492af-39bc-4a73-ae51-3bc39954c358.png)) using the DGkernel.

Could anyone kindly teach me?

Thank you very much!!

Best regards,
Somnus Yu

**Comment by GiudGiud** [comment, marked as answer]
Hello

You can use a regular kernel to express that term. It's a regular source term.
In making your new kernel, you can start from
https://mooseframework.inl.gov/source/kernels/GradientComponent.html
then add a factor, phi, using the coupledValue() routine.
An example using this routine is the MatReaction kernel.

Guillaume

**Reply by SomnusYu** [reply]
Hi,

Thanks for your reply. I know the regular kernel, but there are convection terms in the coupled equations, which are better to use DGkernel. And I try to use DGkernel for convection terms and regular kernel for ![image](https://user-images.githubusercontent.com/118275988/211011788-31cfdf00-dabe-4d25-914b-4f3609cc8675.png) term.

 But I find that the DGkernel can not work simultaneously with the regular kernel, which means if both the two systems are used simultaneously, the regular kernel will be null. So how to deal with this problem?

Thank you very much!!

Best regards

**Reply by GiudGiud** [reply]
Hello

Which term is null?
phi or dtheta/dx ?
Which type of variables are you using? (what is your [Variables] block)

Guillaume

**Reply by SomnusYu** [reply]
Hi,

null means that even if I write this term in the kernel, it doesn't work. It is inactivated. But the terminal still works(does not report an error).
The type of variable is constant MONOMIAL.

Thank you!

**Reply by GiudGiud** [reply]
I dont think it s not active, what makes you say that?
I think the value of the derivative of theta is 0 because the derivative of a constant monomial on an element is 0. 
You ll need to use a different procedure to compute this derivative. 

Is getVar("theta").gradSln() getting you the correct value for the gradient?

Otherwise we ll have to look at some sort of Green-Gauss gradient, and I ll ask others to pitch in

**Reply by SomnusYu** [reply]
Thank you, I will have a try!!

**Reply by SomnusYu** [reply]
Hi,

I find that the gradSln() is only used in FVKernel. Is there any other way to calculate the derivation (variable type is constant monomial) in the kernel system?

Thank you very much!!

**Reply by GiudGiud** [reply]
Kernel.C uses
```
_grad_u(_var.gradSln())
```
to initialize the gradient of the variable.
This then gets used for the ADDiffusion kernel.

**Reply by SomnusYu** [reply]
Hi,

Thanks for your reply. I have used a small case to test if "gradSln()" works. But it shows that the gradient is still 0. Here are my case, the source file and the terminal screenshot.

[test_grad.txt](https://github.com/idaholab/moose/files/10366240/test_grad.txt)
[GradTest.C.txt](https://github.com/idaholab/moose/files/10366241/GradTest.C.txt)
[GradTest.h.txt](https://github.com/idaholab/moose/files/10366242/GradTest.h.txt)
![screenshot](https://user-images.githubusercontent.com/118275988/211150680-508faef3-f7c8-4b7d-bc3c-365f177164da.png)

Thank you!

**Reply by GiudGiud** [reply]
can you make theta a higher order variable? maybe first order

**Reply by SomnusYu** [reply]
Actually, I use the RDG(reconstructed discontinuous Galerkin) model, which needs the variables to be constant MONOMIAL. If the higher order is used, there may be wrong in the cell center. Is there any other way?

**Reply by GiudGiud** [reply]
I thought the point of RDG was to be able to use higher order variables as well but @joshuahansel will know. 

@joshuahansel how do you get element gradient in RDG for a constant monomial?

@lindsayad do we have a green-gauss gradient utility outside of FV?
we could be fine using a finite volume variable here too

**Reply by SomnusYu** [reply]
Hi,

I try to make the fv=true in [Variables] block, but it seems that the fv variable is another system which can not be used in RDG model.
![1673256104753](https://user-images.githubusercontent.com/118275988/211275698-b216b08a-78ed-4077-ad10-0f7990736ed3.png)

So I try to make the gradient just like this in the kernel:
![1673256241782](https://user-images.githubusercontent.com/118275988/211276117-08c066eb-ff46-4d77-b1a2-0a001b4de6eb.png)
But the value of the neighbor element can not be acquired.
![1673256313119](https://user-images.githubusercontent.com/118275988/211276680-a50b72ca-9ddc-4885-9274-484017b09fdb.png)
I don't know the reason. Is this way feasible?

Thank you!!!

**Reply by GiudGiud** [reply]
Hi

You might want to wait on getting replies from the others before venturing on these routes. I m not sure where we are on supporting FV variables in FE. 

I ll explain what's going on in case you still want to try:
1) for finite volume we do not use the coupledXXXX routines as much (you can use adCoupledValue). We use the functor interface most of the time, so the code should look like: getFunctor<ADReal> to initialize the variable.
Functors may be variables, functions or functor material property. You can see examples in fvkernels. 
In FVDiffusion for example, the _coef is initialized by 
```
    _coeff(getFunctor<ADReal>("coeff")),
```
then you can take the gradient of a functor very easily:
`_coef.grad(<arg>)`

`<arg>` should be an element argument. There's routines to generate those, such
`makeElemArg(Elem * elem)`

2) for the finite difference gradient you are trying to make, this will not work as is because the neighbor is not defined for a regular kernel. It is only defined for interface kernels, and other face operators. For a regular kernel, there are more than one neighbor.

Guillaume

**Reply by SomnusYu** [reply]
Hi,

Thanks for your reply and explanation. I will wait for others' reply!!

Thank you!!

**Reply by joshuahansel** [reply]
Hi Somnus,

We have some equations with a non-conservative term like your `phi*dtheta/dx` term, and we're using a FV scheme equivalent to P0P1 RDG. Our variables are CONST MONOMIAL as well. In our scheme, we've made the approximation that your `phi` coefficient on that term is constant over the element, so then after integration over the element you effectively have a flux `phi_i * theta_{i+1/2}` on element i and `phi_{i+1} * theta_{i+1/2}` on element i+1.

**Reply by joshuahansel** [reply]
> I thought the point of RDG was to be able to use higher order variables as well but @joshuahansel will know.

Yes, an RDG implementation should be able to use higher order variables, but as far as I know, we don't have any good way to generalize it to higher order in MOOSE yet (probably doable though).

> @joshuahansel how do you get element gradient in RDG for a constant monomial?

We haven't needed to do this yet (see my previous answer). However, maybe you could use the higher order reconstruction for this term.

**Reply by SomnusYu** [reply]
Hi,

Thanks for your reply. So, you mean there is no way to construct the non-conservative term just using RDG and a regular kernel system? What I need to do is discard my original code and rewrite it in the FV system. Is that right?

Thank you!!

**Reply by joshuahansel** [reply]
No, I didn't mean to suggest that. Here are two options (there could be others):

- Approximate `phi` as constant over the element and use DGKernels to add the resulting` phi_i * theta_{i+1/2}` and `phi_{i+1} * theta_{i+1/2}` fluxes.
- Use your reconstruction scheme to estimate `dtheta/dx` and then add `phi*dtheta/dx` via a (regular) Kernel.

**Reply by SomnusYu** [reply]
Hi Joshuahansel,

Thank you very much! 

For the first option, I have coded it in DGkernel, is that right?
```
  ADReal re = 0.0;
  const ADReal theta_flux = 0.5 * (_theta1[_qp] + _theta2[_qp]); //_theta1 is current element value and _theta2 is neighbor element value.
  switch (type) 
  {
    case Moose::Element:
      re = _phi1[_qp] * theta_flux * _test[_i][_qp];
    case Moose::Neighbor:
      re = - _phi2[_qp] * theta_flux * _test_neighbor[_i][_qp];
    break;
  }
  return re;
```

For the second option, I do not know how to do it.  Could you please tell me more details?

Thank you!!
Best regards

**Reply by joshuahansel** [reply]
Yes, the first option looks correct. For the second, you said that you're using RDG. How are you doing or planning to do your reconstruction (R)? I'm just suggesting that whatever technique you use to build the reconstruction for the fluxes, you also use for this non-conservative term.

**Reply by SomnusYu** [reply]
Thank you!!  For the second option, do you mean I need to use a reconstruction scheme (just like MC, MINMODE, SUPERBEE et al.) to calculate the **dtheta/dx** in Material, and then coupled this material to the regular kernel? Finally, add **phi*dtheta/dx** in regular kernel?

**Reply by joshuahansel** [reply]
Yes, something like that!

**Reply by SomnusYu** [reply]
Thank you!! I will have a try!! Much appreciate!!

**Reply by lindsayad** [reply]
Why do you want to use rDG as opposed to standard DG?

**Reply by SomnusYu** [reply]
Hi lindsayad,

The standard DG is equivalent to a first-order FV scheme, which may cause large numerical diffusion. But rDG can reach second-order numerical precision.

**Reply by SomnusYu** [reply]
Hi joshuahansel,

I have tried the second option, but it seems not correct. I think maybe it is because of the boundary condition. 
I used the freeoutflow boundary condition, just like the rDG model. I need to give the boundary flux of the conservative term (**d(phi*v)/dx**), so the flux of freeoutflowBC should be **phi * v**. Need the non-conservative term (**phi*dtheta/dx**) also be constructed in BC? If so, how to construct it?

Thank you!!

**Reply by joshuahansel** [reply]
> The standard DG is equivalent to a first-order FV scheme, which may cause large numerical diffusion. But rDG can reach second-order numerical precision.

I'm not sure what is meant here. DG is able to use higher-order finite elements and thus achieve greater than first-order accuracy. As I understand it, RDG basically just buys you potentially being able to get to a given order of accuracy more efficiently, but at the cost of the simplicity of standard DG. I think the only question is about whether your scheme is stabilized or not - it may become more difficult to achieve certain stability requirements at higher orders.

> I have tried the second option, but it seems not correct.

I'm not entirely sure that it would work in the first place - it was just an idea. That being said, did your reconstruction use limiting as well? If so, maybe don't use limitation in this term - instead, only use limitation for the conservative fluxes and not your source term. If you're implementing the term as a Kernel, you should not need any BC associated with it. You'd only need an additional BC if you were using DGKernels in the first option.

**Reply by SomnusYu** [reply]
> it may become more difficult to achieve certain stability requirements at higher orders.

Yes, stability is really a problem using higher orders, so I just use the variable type of constant monomial.


> If so, maybe don't use limitation in this term - instead, only use limitation for the conservative fluxes and not your source term.

Thank you for your suggestion, I will have a try.

> If you're implementing the term as a Kernel, you should not need any BC associated with it. You'd only need an additional BC if you were using DGKernels in the first option.

Thank you!
